### PR TITLE
Fix and refactor NumberTraits

### PR DESCRIPTION
**Pull request by rolanddenis**
# PR Description

Fix issues with `NumberTraits` (see #1396) and refactor this class by relying on `std::numeric_limits` when possible (i.e. for all fundamental arithmetical types).

Also add a test file for NumberTraits.

# Checklist

- [x] Unit-test of your feature with [Catch](http://dgtal.org/doc/stable/moduleCatch.html).
- [X] Doxygen documentation of the code completed (classes, methods, types, members...)
- [x] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [x] No warning raised in Debug ```cmake``` mode (otherwise, Travis C.I. will fail).
- [x] All continuous integration tests pass (Travis & appveyor)

### CODE REVIEW

**Comment by dcoeurjo** [comment]
thanks @rolanddenis .. reviewing ;)

File: tests/kernel/testNumberTraits.cpp [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+  using type = DGtal::TagTrue;
+};
+
+template <>
+struct ValueToTag<false>
+{
+  using type = DGtal::TagFalse;
+};
+
+#define REQUIRE_SAME_TAG(a, b)    REQUIRE( (std::is_same<a, typename IntegralCstToTag<b>::type>::value) )
```
Comment (dcoeurjo) [inline tests/kernel/testNumberTraits.cpp]: nice macros;)

Comment (rolanddenis) [reply tests/kernel/testNumberTraits.cpp]: :sweat_smile:

Comment (rolanddenis) [reply tests/kernel/testNumberTraits.cpp]: I realize that I don't even use this macro...

**Comment by dcoeurjo** [comment]
well you're way more skilled than me for this kind of type related stuffs...

File: tests/kernel/testNumberTraits.cpp (hunk 1 above)
Comment (dcoeurjo) [reply tests/kernel/testNumberTraits.cpp]: ^^

**Comment by dcoeurjo** [comment]
Let's merge this one;) thx

**Pull request merged**